src/pie_utils/sequence_tagging/encoding.py

```diff
@@ -0,0 +1,363 @@
+    classes_to_ignore = classes_to_ignore or []
+    index = 0
+    while index < len(tag_sequence):
+        label = tag_sequence[index]
+        if label[0] == "B":
+            start = index
+            current_span_label = label.partition("-")[2]
+            index += 1
+            if index < len(tag_sequence):
+                label = tag_sequence[index]
+            else:
+                # if B is last tag in the sequence
+                spans.append((current_span_label, (start, start)))
+                continue
+            if label[0] == "B":
+                # if Ba Bb or Ba Ba
+                spans.append((current_span_label, (start, start)))
+                continue
+            while label[0] == "I" and index < len(tag_sequence):
+                # loop can end if it encounters another B or O or end of tag_sequence
+                index += 1
+                if index < len(tag_sequence):
+                    label = tag_sequence[index]
+            index -= 1
+            end = index
+            spans.append((current_span_label, (start, end)))
+        index += 1
+    return [span for span in spans if span[0] not in classes_to_ignore]
+
+
+def bioul_tags_to_spans(
+    tag_sequence: List[str],
+    classes_to_ignore: List[str] = None,
+) -> List[TypedStringSpan]:
+    """Given a sequence corresponding to BIOUL tags, extracts spans. Spans are inclusive and can be
+    of zero length, representing a single word span. Ill-formed spans are not allowed and will
+    raise `InvalidTagSequence`. This function works properly when the spans are unlabeled (i.e.,
+    your labels are simply "B", "I", "O", "U", and "L").
+
+    # Parameters
+
+    tag_sequence : `List[str]`, required.
+        The tag sequence encoded in BIOUL, e.g. ["B-PER", "L-PER", "O"].
+    classes_to_ignore : `List[str]`, optional (default = `None`).
+        A list of string class labels `excluding` the bio tag
+        which should be ignored when extracting spans.
+
+    # Returns
+
+    spans : `List[TypedStringSpan]`
+        The typed, extracted spans from the sequence, in the format (label, (span_start, span_end)).
+    """
+
+    spans = []
+    classes_to_ignore = classes_to_ignore or []
+    index = 0
+    while index < len(tag_sequence):
+        label = tag_sequence[index]
+        if label[0] == "U":
+            spans.append((label.partition("-")[2], (index, index)))
+        elif label[0] == "B":
+            start = index
+            end = index
+            current_span_label = label.partition("-")[2]
+            while label[0] != "L" and index < len(tag_sequence):
+                index += 1
+                label = tag_sequence[index]
+                if label[0] == "L":
+                    end = index
+            spans.append((current_span_label, (start, end)))
+        index += 1
+    return [span for span in spans if span[0] not in classes_to_ignore]
+
+
+def _to_bioul(tag_sequence: List[str], encoding: str = "IOB1") -> List[str]:
```

Comment:
* is `encoding` used in any way? remove? 
* maybe renaming to `_iob2_to_bioul`? 
* is the input `IOB2` or `IOB1`? this should be reflected in the doc string!

Reply:
Yes, encoding is not used anymore since we always expect encoding to be `IOB2`. Regarding renaming then I'll rename all internal function to start with `_`.